### DISCUSSION TITLE
Primary + unique constraint on same migration

### DISCUSSION THREAD

**Discussion by Gildraen** [sqlalchemy/alembic#1494] [Usage Questions]
**Describe the bug**
Hi, when i want to create a table with a column that has primary AND unique:

` id: MappedColumn[UUID] = mapped_column(
        UserIdType(as_uuid=True), primary_key=True, unique=True, nullable=False
    )`

the migration generated seems good:
`op.create_table(
        "users",
        sa.Column("id", UserIdType(), nullable=False),
        sa.PrimaryKeyConstraint("id"),
        sa.UniqueConstraint("id"),
    ) `
but it does not create the uniq index.. i only have the primary index

**Expected behavior**
The creation of the table, with the primary index and the unique index on the same migration

**To Reproduce**

```py
# Insert code here
from uuid import UUID
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy import Uuid
from sqlalchemy.orm import MappedColumn
from sqlalchemy.orm import mapped_column

class User(DeclarativeBase):
    __tablename__ = "users"

    id: MappedColumn[UUID] = mapped_column(
        Uuid(as_uuid=True), primary_key=True, unique=True, nullable=False
    )
```

**Versions.**
 - OS: Debian 12
 - Python: 3.11.9
 - Alembic:1.13.1
 - SQLAlchemy:2.0.29
 - Database: Postgresql:16
 - DBAPI:psycopg2

**Have a nice day!**

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you are asking here. The generated migration is correct, so I'm not sure what you are look for

**Comment by Gildraen** [comment]
Hi, totally, the migration is correct, but the final result is not

I don't know if it's related to alembic or to postgresql

but the result is : there is no unique key constraint on the table, only the primary key

how can i see the raw sql generated from alembic ? maybe it could help for understanding ?

Here is the result of the migration as mentionned on op (to show that there is one constraint missing)
![image](https://github.com/sqlalchemy/alembic/assets/20871857/4a8064da-81ef-4f0f-bb52-678cd0a966d0)

**Reply by CaselIT** [reply]
you can set the echo flag on the sqlalchemy engine.

note that it makes little sense to define both a pk and an unique constraint on the same column, but I don't think pg will prevent you from doing that. also note that an unique constraint is not an unique index so it may not be listed under indexes in your ui

**Reply by Gildraen** [reply]
For the sense, i am aware of that, and i do not "need" that actually, 
But they are not exclusive, and are independent, and in my case, i like to have explicit things, so having both is not necessary, but still should be possible
it's not a problem of ui because : 
 - I can create it from the ui or via sql without problem
 - Alembic itself want to create the unique index on the next migration  (so it see that there is something missing)

after showing sql, thank you for that, i can see that the sql generated seems good : 

`CREATE TABLE users (
    id UUID NOT NULL,
    PRIMARY KEY (id),
    UNIQUE (id)
);
`

and i tried directly on the db, and it does the same : no unique index. so maybe is postgresql related....

( i know i can have a workaround with this : `op.create_unique_constraint(
        None, "users", ["id"]
    ) after the first creation)`

**Reply by CaselIT** [reply]
I guess that when present at the same time in the create table statement pg does in fact ignore the unique constraint, since that's implied by the primary key.
Seems to be a pg specific behaviour. This use case seems rare enough that I don't think it's worth the effort to support. You can manually add the unique constraint as you mentioned

**Reply by Gildraen** [reply]
Based on that, is it usefull, intended, that alembic generate these migration with both constraint ? 
Like, creating it, and let to the db the decision of what to do with it ? 
or maybe there is some scenario where both would be taken in consideration even in pg ?

**Reply by CaselIT** [reply]
There is really no logic here regarding selecting what to render.
The migration has both so both are rendered

**Reply by Gildraen** [reply]
Okey, and it seems relevant for me, not the job of Alembic do determine those specifics choices (as long as it doesn't make the migration break ^^)

**Reply by CaselIT** [reply]
it's actually quite strange that pg decides here to not do something. usually is happy to let the user do what it pleases even if it does not make much sense or it's not needed. like having two indexes in the same column etc.

not sure why they decided to handle this case differently, considering that it then lets you create an unique constraint if you use alter table

**Reply by Gildraen** [reply]
Exactly, i still can add the constraint on a second hand, so yea, that's what i finded strange.. i may ask on their repo, who knows ! Maybe is not intended
